sdk/spring/azure-spring-integration-core/src/main/java/com/azure/spring/integration/core/DefaultMessageHandler.java

```diff
@@ -163,22 +165,40 @@ private String toDestination(Message<?> message) {
 
     private PartitionSupplier toPartitionSupplier(Message<?> message) {
         PartitionSupplier partitionSupplier = new PartitionSupplier();
-        String partitionKey = message.getHeaders().get(AzureHeaders.PARTITION_KEY, String.class);
-        if (!StringUtils.hasText(partitionKey) && this.partitionKeyExpression != null) {
-            partitionKey = this.partitionKeyExpression.getValue(this.evaluationContext, message, String.class);
+        // Priority setting partitionId
+        String partitionId = getHeaderValue(message.getHeaders(), AzureHeaders.PARTITION_ID);
+        if (!StringUtils.hasText(partitionId) && this.partitionIdExpression != null) {
+            partitionId = this.partitionIdExpression.getValue(this.evaluationContext, message, String.class);
         }
-
-        if (StringUtils.hasText(partitionKey)) {
-            partitionSupplier.setPartitionKey(partitionKey);
-        }
-
-        if (message.getHeaders().containsKey(AzureHeaders.PARTITION_ID)) {
-            partitionSupplier
-                .setPartitionId(message.getHeaders().get(AzureHeaders.PARTITION_ID, String.class));
+        if (StringUtils.hasText(partitionId)) {
+            partitionSupplier.setPartitionId(partitionId);
+        } else {
+            String partitionKey = getHeaderValue(message.getHeaders(), AzureHeaders.PARTITION_KEY);
+            // The default key expression is the hash code of the payload.
+            if (!StringUtils.hasText(partitionKey) && this.partitionKeyExpression != null) {
+                partitionKey = this.partitionKeyExpression.getValue(this.evaluationContext, message, String.class);
+            }
+            if (StringUtils.hasText(partitionKey)) {
+                partitionSupplier.setPartitionKey(partitionKey);
+            }
         }
         return partitionSupplier;
     }
 
+    /**
+     * Get header value from MessageHeaders
+     * @param headers MessageHeaders
+     * @param keyName Key name
+     * @return String header value
+     */
+    private String getHeaderValue(MessageHeaders headers, String keyName) {
+        return headers.keySet().stream()
+            .filter(header -> keyName.equals(header))
+            .map(key -> String.valueOf(headers.get(key)))
+            .findAny()
+            .orElse(null);
+    }
+
```

Comment:
```
    /**
     * Get header value from MessageHeaders
     *
     * @param message MessageHeaders
     * @param keyName Key name
     * @return String header value
     */
    private String getHeaderValue(Message<?> message, String keyName) {
        return (String) Optional.ofNullable(message)
                                .map(Message::getHeaders)
                                .map(headers -> headers.getOrDefault(keyName, null))
                                .orElse(null);
    }

    private Map<String, Object> buildPropertiesMap() {
        Map<String, Object> properties = new HashMap<>();
        properties.put("sync", sync);
        properties.put("sendTimeout", sendTimeoutExpression);
        properties.put("destination", destination);

        return properties;
    }
```
1. Avoid NullPointerException(NPE).
2. Avoid visit map 2 times: `keyName.equals(header)` and `headers.get(key)`.